Start Entities Prototype

tigrannajaryan opened the pull request:
Changes:
- Add new Entity signal (EntityEvents) according to OTEP 0256.
- Add EntityRefs to Resource according to OTEP 0264.
- Correctly indicate unstable elements

tigrannajaryan committed:
Add entities

tigrannajaryan committed:
Fix field names to make them Collector pdata friendly

tigrannajaryan committed:
Update proto to reflect OTEP 0264 requirements

There is now multiple ResourceEntityRefs in the Resource.

smith commented on opentelemetry/proto/collector/entities/v1/entities_service.proto:
```diff
@@ -0,0 +1,75 @@
+option go_package = "go.opentelemetry.io/proto/otlp/collector/entities/v1";
+
+// Service that can be used to push entities between one Application instrumented with
+// OpenTelemetry and an collector, or between an collector and a central collector (in this
```
```suggestion
// OpenTelemetry and a collector, or between a collector and a central collector (in this
```

tigrannajaryan replied:
Fixed.

smith commented on opentelemetry/proto/collector/entities/v1/entities_service.proto:
```diff
@@ -0,0 +1,75 @@
+}
+
+message ExportEntitiesPartialSuccess {
+  // The number of rejected log records.
```
Is this supposed to say "log records"?

tigrannajaryan replied:
Good catch. Copy/paste error. Fixed now.

smith commented on opentelemetry/proto/entities/v1/entities.proto:
```diff
@@ -0,0 +1,82 @@
+option java_outer_classname = "EntitiesProto";
+option go_package = "go.opentelemetry.io/proto/otlp/entities/v1";
+
+// EntitiesData represents the entities data that can be stored in a persistent storage,
```
```suggestion
// EntitiesData represents the entities data that can be stored in persistent storage,
```

tigrannajaryan replied:
Fixed.

tigrannajaryan committed:
Address PR comments

tigrannajaryan committed:
Correctly indicate unstable elements

- Match the requirements defined in https://github.com/open-telemetry/opentelemetry-proto/pull/597
- Use v1development instead of v1 to match what we do for profiles: https://github.com/open-telemetry/opentelemetry-proto/tree/main/opentelemetry/proto/profiles/v1development

jsuereth commented on opentelemetry/proto/resource/v1/resource.proto:
```diff
@@ -34,4 +34,39 @@ message Resource {
+  // Attribute Keys that identify the entity.
+  // MUST not change during the lifetime of the entity. The Id must contain at least one attribute.
+  // These keys MUST exist in the containing Resource.attributes.
+  repeated string id_attr_keys = 3;
```
Adding this, but don't want to start a huge bikeshedding argument.

Let's be confident we are ok with these names in json.

Alternative suggestions (in order of my preference):

- `id_keys`, `desc_keys`
- `id_keys`, `descr_keys`
- `identifying_keys`, `descriptive_keys`
- `id_attr_keys`, `desc_attr_keys`
- `id_refs`, `desc_refs`

feldentm-SAP replied:
TBH, I still do not think that having a schema and a set of keys is a good idea; I would expect that having a schema only and baking both key sets into the code is a far better solution; this is data that we would ship over the network all the time

jsuereth reviewed (approved):
LGTM - but one question on expected JSON.

feldentm-SAP commented on opentelemetry/proto/entities/v1development/entities.proto:
```diff
@@ -0,0 +1,82 @@
+  // The instrumentation scope information for the entities in this message.
+  // Semantically when InstrumentationScope isn't set, it is equivalent with
+  // an empty instrumentation scope name (unknown).
+  opentelemetry.proto.common.v1.InstrumentationScope scope = 1;
```
What is an instrumentation scope in this context? I would have expected entities to be orthogonal to instrumentation.

feldentm-SAP commented on opentelemetry/proto/collector/entities/v1development/entities_service.proto:
```diff
@@ -0,0 +1,75 @@
+option go_package = "go.opentelemetry.io/proto/otlp/collector/entities/v1development";
+
+// Service that can be used to push entities between one Application instrumented with
+// OpenTelemetry and a collector, or between a collector and a central collector (in this
```
wouldn't that imply that collectors loose their stateless scalability property? that change would imply that collectors need to receive all entity event to deliver entity-based enrichment

feldentm-SAP commented on opentelemetry/proto/collector/entities/v1development/entities_service.proto:
```diff
@@ -0,0 +1,75 @@
+
+// Service that can be used to push entities between one Application instrumented with
+// OpenTelemetry and a collector, or between a collector and a central collector (in this
+// case entities are sent/received to/from multiple Applications).
```
what happens if multiple applications send partial or conflicting information?
